### PR TITLE
[kie-issues#745] Handle lists and maps in method calls and constructors in drools-mvel-compiler

### LINKED ISSUE #745: Handle lists and maps in method calls and constructors in drools-mvel-compiler
Currently it is not possible to use [] mvel construct with Drools executable model as it is not handled in drools-mvel-compiler. The construct is parsed in mvel parser. 

Acceptance criteria: 
- Handle map and list mvel classes produced by drools-mvel-parser in drools-mvel-compiler.
- Generate Drools Java executable model code in drools-mvel-compiler based on the map and list classes provided by drools-mvel-parser.

### PR DESCRIPTION
**Pull request by baldimir**
Fixes https://github.com/apache/incubator-kie-issues/issues/745. 

- Adds maps and lists handling to drools-mvel-compiler. 
- Adds a collection type coercion for mvel empty list and map expression ambiguity ([] can be an empty list but also an empty map).

### CODE REVIEW

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/MethodCallExprVisitor.java [hunk 1]
```diff
@@ -89,43 +87,21 @@ private MethodCallExprT parseMethod(MethodCallExpr n,
...
-                    .map(scopeClazz -> MethodUtils.findMethod(scopeClazz, n.getNameAsString(), argumentsType));
-            if(method.isPresent()) {
-                scope = mvelCompilerContext.createRootTypePrefix();
-            }
+                                        List<Integer> emptyCollectionArgumentIndexes) {
+        Pair<Optional<Method>, Optional<TypedExpression>> resolveMethodResult =
+                MethodResolutionUtils.resolveMethod(n, mvelCompilerContext, scope, arguments);
+        // This is a workaround for mvel empty list and map ambiguity, please see the description in getTypedArguments() method.
+        if (resolveMethodResult.a.isEmpty() && !emptyCollectionArgumentIndexes.isEmpty()) {
+            resolveMethodResult = MethodResolutionUtils.resolveMethodWithEmptyCollectionArguments(n, mvelCompilerContext, resolveMethodResult.b, arguments, emptyCollectionArgumentIndexes);
```
Comment (lucamolteni) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/MethodCallExprVisitor.java:95]: Can you please add unit tests in `drools-mvel-compiler` module covering this case?

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/util/MethodResolutionUtils.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+     *    - This happens for all empty list and map parameters resolved by the parser, until a proper constructor is found.
+     */
+    public static List<TypedExpression> coerceCorrectConstructorArguments(
+            final Class<?> type,
+            List<TypedExpression> arguments,
+            List<Integer> emptyListArgumentIndexes) {
+        // Rather work only with the argumentsType and when a method is resolved, flip the arguments list based on it.
+        final List<TypedExpression> coercedArgumentsTypesList = new ArrayList<>(arguments);
+        // This needs to go through all possible combinations.
+        final int indexesListSize = emptyListArgumentIndexes.size();
```
Comment (lucamolteni) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/util/MethodResolutionUtils.java]: Add unit tests for this method as well

**Review by lucamolteni — CHANGES_REQUESTED** [review]

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/MethodCallExprVisitor.java (hunk 1 above)
Comment (baldimir) [reply drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/MethodCallExprVisitor.java:95]: Added.

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/util/MethodResolutionUtils.java (hunk 2 above)
Comment (baldimir) [reply drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/util/MethodResolutionUtils.java]: Added.

**Review by lucamolteni — APPROVED** [review]

**Review by gitgabrio — APPROVED** [review]

**Pull request merged**